torch/csrc/distributed/c10d/ProcessGroup.hpp

```diff
@@ -48,6 +48,77 @@ namespace c10d {
 // process group to find each other (referred to as rendezvous from
 // hereon)
 //
+// We share dimensionality about the sparse tensors before collecting
+// their contents. We assume here that the maximum number of sparse
+// and dense dimensions is 4 each. This is stored in a contiguous piece of
+// memory so that we can easily run allgather on it.
+//
+// The layout of this memory is as follows:
+//
+//   - [0:4]: sparse dims
+//   - [4:8]: dense dims
+//   -   [8]: nnz
+//
+class SparseTensorMetadata {
+ public:
+  static constexpr auto dim = 9;
+
+  // Construct from an existing metadata tensor to facilitate structured
+  // access to metadata from peers, after gathering it.
+  explicit SparseTensorMetadata(at::Tensor metadata)
+      : metadata_(metadata), data_(metadata_.data_ptr<int64_t>()) {
+    AT_ASSERT(metadata.scalar_type() == at::kLong);
+    AT_ASSERT(metadata.dim() == 1);
+    AT_ASSERT(metadata.size(0) == dim);
+  }
+
+  // Populate the metadata.
+  void populate_from_sparse_tensor(const at::Tensor& tensor) {
+    const auto sparse_dim = tensor.sparse_dim();
+    AT_ASSERT(sparse_dim <= 4);
+    for (const auto i : c10::irange(4)) {
+      if (i < sparse_dim) {
+        data_[i] = tensor.size(i);
+      }
+    }
+    const auto dense_dim = tensor.dense_dim();
+    AT_ASSERT(dense_dim <= 4);
+    for (const auto i : c10::irange(4)) {
+      if (i < dense_dim) {
+        data_[i + 4] = tensor.size(sparse_dim + i);
+      }
+    }
+    data_[8] = tensor._nnz();
+  }
+
+  std::vector<int64_t> sizes() const {
+    std::vector<int64_t> sizes;
+    // Sparse sizes
+    for (const auto i : c10::irange(4)) {
+      if (data_[i] <= 0) {
+        break;
+      }
+      sizes.push_back(data_[i]);
+    }
+    // Dense sizes
+    for (const auto i : c10::irange(4, 8)) {
+      if (data_[i] <= 0) {
+        break;
+      }
+      sizes.push_back(data_[i]);
+    }
+    return sizes;
+  }
+
+  int64_t nnz() const {
+    return data_[8];
+  }
+
+ protected:
+  at::Tensor metadata_;
+  int64_t* data_;
+};
+
```

Comment:
I think it would be better if this whole piece of code reside in a separate file than ProcessGroup.hpp.
For example, you can create a SparseTensorMetadata.hpp, and include it where the class SparseTensorMetadata is used.

Reply:
And this piece of code used some numbers like "4" and "8" in code. Defining macros to represent these numbers would make the code more maintainable.

Reply:
sounds good!

Reply:
An alternative to #define macro for "4" and "8" is to use class templates. I think that would make this class more extensible.

For example:
```
template<int SPARSE_DIM, int DENSE_DIM, int NNZ_DIM> 
class SparseTensorMetadata {
  static constexpr auto dim = SPARSE_DIM + DENSE_DIM + NNZ_DIM;
  ...
};

SparseTensorMetadata<4, 4, 1> tensor_meta(...);
```

Reply:
good suggestion!

Reply:
when I think about it more, feel '#define macro' is better for c10d libraries, as it is rare for c10d libraries to support different types of SparseTensorMetadata with different SPARSE_DIM